// smartOpts/smartOpts.js
let smartOpts = function () {
    this.optIn = function (option, list = this.optionsMutable, obj = false, key=false) {
        if (!obj && list.indexOf(option) >= 0) {
            return true
        } else if (obj) {
            for (var i = 0; i < list.length; i++) {
                if (list[i][key] == option[key]) {
                    return true
                }
            }
            return false
        } else {
            return false
        }
    }
    this.optsIn = function (options, list = this.optionsMutable, obj = false, key=false) {
        for (let option of options) {
            console.log(option)
            if (!obj && list.indexOf(option) >= 0) {
                // return true
            } else if (obj) {
                for (var i = 0; i < list.length; i++) {
                    if (list[i][key] == option[key]) {
                        // return true
                        break
                    }
                    if (i == list.length - 1) {
                        return false
                    }
                }
            } else {
                return false
            }
        }
        return true
    }
    this.anyOptsIn = function (options, list = this.optionsMutable, obj = false, key=false) {
        for (let option of options) {
            console.log(option)
            if (!obj && list.indexOf(option) >= 0) {
                return true
            } else if (obj) {
                for (var i = 0; i < list.length; i++) {
                    if (list[i][key] == option[key]) {
                        return true
                    }
                }
            }
        }
        return false
    }
    this.optIndex = function (option, list = this.optionsMutable, obj = false, key=false) {
        if(obj && key){
            for (var i = 0; i < list.length; i++) {
                if (list[i][key] == option[key]) {
                    return i
                }
            }  
        } else {
            return list.indexOf(option)
        }
    }
    this.pushOpt = function (option, list = this.optionsMutable, obj = false, key=false) {
        if (!this.optIn(option, list, obj, key)) {
            list.push(option)
        }
    }
    this.pushOpts = function (options, list = this.optionsMutable, obj = false, key=false) {
        for (let option of options) {
            this.pushOpt(option, list, obj, key)
        }
    }
    this.popOpt = function (option, list = this.optionsMutable, obj = false, key=false) {
        if (this.optIn(option, list, obj, key)) {
            list.splice(this.optIndex(option, list, obj, key), 1)
        }
    }
    this.popOpts = function (options, list = this.optionsMutable, obj = false, key=false) {
        for (let option of options) {
            this.popOpt(option, list, obj, key)
        }
    }
    this.toggleOpt = function (option, list = this.optionsMutable, obj = false, key=false) {
        if (this.optIn(option, list, obj, key)) {
            this.popOpt(option, list, obj, key)
        } else {
            this.pushOpt(option, list, obj, key)
        }
    }
    this.toggleOpts = function (options, list = this.optionsMutable, obj, key) {
        for (let option in options) {
            if (this.optIn(option, list, obj, key)) {
                this.popOpt(option, list, obj, key)
            } else {
                this.pushOpt(option, list, obj, key)
            }
        }
    }
}

module.exports = new smartOpts
